### PR TITLE
Magnet Drop ALL the things!

### PR DESCRIPTION
**Pull request by DingleFlop**
[Changelogs]: # 
:cl:
add: Datum component for magnet drop functionality.
refactor: Refactor cyberimp arm implants to use new magnet drop component.
/:cl:

[why]: # (Please add a short description [two lines down] of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.)

I like how the janitor's cleaner backpack and the defibrillator work, and I think it would be nicer/more consistent if more things had that behaviour. This adds an easy way for items to have that functionality without rewriting code.

### CODE REVIEW

**Comment by DingleFlop** [comment]
First PR, please point out things I could do better, as I am still getting familiar with /tg/ coderbus.

File: code/modules/surgery/organs/augments_arms.dm [hunk 1]
```diff
@@ -20,7 +22,16 @@
 
 	update_icon()
 	SetSlotFromZone()
-	items_list = contents.Copy()
+	create_items(normal_items)
+
+/obj/item/organ/cyberimp/arm/proc/create_items(var/list/arg_items, var/emag = 0)
+	for(var/listitem in arg_items)
+		if(ispath(listitem) && !(locate(listitem) in items_list))
```
Comment (ninjanomnom) [inline code/modules/surgery/organs/augments_arms.dm]: Just let it error if something isn't a path. In the same vein let it add a second of the same item.

File: code/modules/surgery/organs/augments_arms.dm [hunk 2]
```diff
@@ -20,7 +22,16 @@
...
+			var/obj/item/I = new listitem(src)
+			I.AddComponent(/datum/component/magnet_drop, src)
+			if(emag)
+				to_chat(usr, "<span class='notice'>You unlock [src]'s integrated [I.name]!</span>")
```
Comment (ninjanomnom) [inline code/modules/surgery/organs/augments_arms.dm]: `usr` isn't allowed in most cases outside verbs. You'll need to get the user through some other method, probably by just seeing who's wearing the thing.

(The older code is an example of code being untouched for ages)

File: code/modules/surgery/organs/augments_arms.dm [hunk 3]
```diff
@@ -84,7 +95,7 @@
 
 	holder = item
 
-	holder.flags_1 |= NODROP_1
+	//holder.flags_1 |= NODROP_1
```
Comment (ninjanomnom) [inline code/modules/surgery/organs/augments_arms.dm]: just remove it

File: code/modules/surgery/organs/augments_arms.dm [hunk 4]
```diff
@@ -146,12 +157,14 @@
 	else // The gun will still discharge anyway.
 		..()
 
+/obj/item/organ/cyberimp/arm/emag_act()
+	create_items(emag_items, 1)
```
Comment (ninjanomnom) [inline code/modules/surgery/organs/augments_arms.dm]: `TRUE` instead of `1`

**Comment by kevinz000** [comment]
needs a proccall/callback for when the component detects that it was dropped so things can handle putting said object back into it, forcemove is not enough in the vast majority of cases to my knowledge

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/modules/surgery/organs/augments_arms.dm (hunk 2 above)
Comment (DingleFlop) [reply code/modules/surgery/organs/augments_arms.dm]: Got it!

Comment (DingleFlop) [reply code/modules/surgery/organs/augments_arms.dm]: "owner" seems to be a good way to fix this.

**Comment by DingleFlop** [comment]
@kevinz000 This is probably true, I am not sure the best way to go about this, though.

**Comment by DingleFlop** [comment]
In fact, I just noticed a semi-issue; If you click the implant action button, it alerts you that the item is retracted, while the if you drop it, it just disappears without alerting the user.

Could either do a proc callback, or make the component add sound and a chat message as an option, which I am pretty sure goes against the whole point of components.

**Comment by ninjanomnom** [comment]
You probably want to have the action as part of the component itself in that case

**Comment by kevinz000** [comment]
shoudl make a proc callback and let the thing that has it handle the rest. have it call a callback if it exists when the magnet dropping triggers.
also, please make sure that this checks if the object is a storage. if it is, it should be using a forceful storage insertion signal instead of a direct forcemove.
infact @ninjanomnom should i just make forcemoves into storage trigger a storage forced insertion if it isn't from the storage already? that's the other option, as one of the two has to be done or it'll break things when storage is involved.

**Comment by DingleFlop** [comment]
I am still working on this. I will tell you when I am ready for another pass.

File: code/modules/projectiles/guns/energy/mounted.dm [hunk 5]
```diff
@@ -21,6 +18,3 @@
 	force = 5
 	selfcharge = 1
 	trigger_guard = TRIGGER_GUARD_ALLOW_ALL
-
```
[the remark this answers is not in the document]
Comment (ShizCalev) [reply code/modules/projectiles/guns/energy/mounted.dm]: No reason to have them.

**Comment by optimumtact** [comment]
sir

**Comment by DingleFlop** [comment]
Gunna finish this up this weekend, sorry for letting it simmer.

**Comment by ninjanomnom** [comment]
Let me know when you've started work on this again, I was interested in this.

**Pull request closed without merging**